src/Bedrock.Framework.Experimental/Transports/WebSockets/WebSocketFrameWriter.cs

```diff
@@ -0,0 +1,114 @@
+        /// <param name="output">The buffer writer to write the message to.</param>
+        public unsafe void WriteMessage(ref WebSocketWriteFrame message, IBufferWriter<byte> output)
+        {
+            if(message.Header.PayloadLength != (ulong)message.Payload.Length)
+            {
+                throw new WebSocketFrameException($"Header payload length ({message.Header.PayloadLength}) does not equal supplied payload length ({message.Payload.Length})");
+            }
+
+            var headerSize = 2;
+            var extendedPayloadLengthSize = 0;
+            var maskPosition = 2;
+
+            if(message.Header.PayloadLength > 125)
+            {
+                if(message.Header.PayloadLength <= ushort.MaxValue)
+                {
+                    extendedPayloadLengthSize = 2;
+                    maskPosition += 2;
+                }
+                else
+                {
+                    extendedPayloadLengthSize = 8;
+                    maskPosition += 8;
+                }
+            }
+
+            if(message.Header.Masked)
+            {
+                headerSize = headerSize + extendedPayloadLengthSize + 4;
+            }
+            else
+            {
+                headerSize = headerSize + extendedPayloadLengthSize;
+            }
+
+            Span<byte> headerSpan = stackalloc byte[headerSize];
+            headerSpan[0] = (byte)message.Header.Opcode;
+
+            if(message.Header.Fin)
+            {
+                headerSpan[0] |= 0b1000_0000;
+            }
+
+            switch(extendedPayloadLengthSize)
+            {
+                case 2:
+                    headerSpan[1] = 126;
+                    headerSpan[2] = (byte)(message.Header.PayloadLength >> 8);
+                    headerSpan[3] = unchecked((byte)message.Header.PayloadLength);
+                    break;
+                case 8:
+                    headerSpan[1] = 127;
+                    ulong length = message.Header.PayloadLength;
+                    for (int i = 9; i >= 2; i--)
+                    {
+                        headerSpan[i] = unchecked((byte)length);
+                        length = length >> 8;
+                    }
+                    break;
+                default:
+                    headerSpan[1] = (byte)message.Header.PayloadLength;
+                    break;
+            }
+
+            if(message.Header.Masked)
+            {
+                headerSpan[1] |= 0b1000_0000;
+                headerSpan[maskPosition] = (byte)message.Header.MaskingKey;
+                headerSpan[maskPosition + 1] = (byte)(message.Header.MaskingKey >> 8);
+                headerSpan[maskPosition + 2] = (byte)(message.Header.MaskingKey >> 16);
+                headerSpan[maskPosition + 3] = (byte)(message.Header.MaskingKey >> 24);
+
+                if(!message.MaskingComplete)
+                {
+                    var useSimd = Vector.IsHardwareAccelerated && Vector<byte>.Count % sizeof(int) == 0;
```

Comment:
The entirety of the unmasking code was more or less lifted from ManagedWebSocket in corefx. I tried to clarify some of it a touch though.

My assumption of the intention of that code is in case this runs on a architecture where that would not be the case. I don't know of any such architectures off the top of my head, though.

Reply:
> runs on a architecture where that would not be the case.

Ah, this sounds reasonable.
I don't know any such architecture, too. But there may be more than x86 and ARM...which are the architectures where `Vector` is actually hardware accelerated in the moment. So in reality this will always result in true.
As there is an tendency to implement Vector on top of HW-intrinsics I strongly assume this will be true in the future too.
In effect this check can be removed safely.

And keep my other comments about `useSimd` in mind :wink:

Reply:
You've convinced me for now. This is easy enough code to add back if other archs get hardware accelerated Vector support. I've removed the useSimd usage as well.